[CI] RestCancellableNodeClientTests testConcurrentExecuteAndClose failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #8202 / ubuntu-2004-aarch64_checkpart1_platform-support-arm](https://gradle-enterprise.elastic.co/s/a5u67t43ckdcu)
- [elasticsearch-periodic-platform-support #8094 / ubuntu-2004_platform-support-unix](https://gradle-enterprise.elastic.co/s/g34et3cugdpce)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.rest.action.RestCancellableNodeClientTests.testConcurrentExecuteAndClose" -Dtests.seed=C00B660EC9AB68A2 -Dtests.locale=kok-Deva -Dtests.timezone=Europe/Zagreb -Druntime.java=24
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.rest.action.RestCancellableNodeClientTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testConcurrentExecuteAndClose),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected:<0> but was:<1>
```

**Issue Reasons:**
- [main] 2 failures in test testConcurrentExecuteAndClose (0.3% fail rate in 717 executions)
- [main] 2 failures in pipeline elasticsearch-periodic-platform-support (10.5% fail rate in 19 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/server/src/test/java/org/elasticsearch/rest/action/RestCancellableNodeClientTests.java
+++ b/server/src/test/java/org/elasticsearch/rest/action/RestCancellableNodeClientTests.java
@@ -19,6 +19,7 @@ import org.elasticsearch.action.search.SearchRequest;
 import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.action.search.TransportSearchAction;
 import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.action.support.SubscribableListener;
 import org.elasticsearch.client.internal.node.NodeClient;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.set.Sets;
@@ -44,7 +45,6 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.LongSupplier;
 
 public class RestCancellableNodeClientTests extends ESTestCase {
@@ -79,7 +79,9 @@ public class RestCancellableNodeClientTests extends ESTestCase {
             for (int j = 0; j < numTasks; j++) {
                 PlainActionFuture<SearchResponse> actionFuture = new PlainActionFuture<>();
                 RestCancellableNodeClient client = new RestCancellableNodeClient(testClient, channel);
-                threadPool.generic().submit(() -> client.execute(TransportSearchAction.TYPE, new SearchRequest(), actionFuture));
+                futures.add(
+                    threadPool.generic().submit(() -> client.execute(TransportSearchAction.TYPE, new SearchRequest(), actionFuture))
+                );
                 futures.add(actionFuture);
             }
         }
@@ -150,7 +152,7 @@ public class RestCancellableNodeClientTests extends ESTestCase {
         assertEquals(totalSearches, testClient.cancelledTasks.size());
     }
 
-    public void testConcurrentExecuteAndClose() throws Exception {
+    public void testConcurrentExecuteAndClose() {
         final var testClient = new TestClient(Settings.EMPTY, threadPool, true);
         int initialHttpChannels = RestCancellableNodeClient.getNumChannels();
         int numTasks = randomIntBetween(1, 30);
@@ -254,7 +256,7 @@ public class RestCancellableNodeClientTests extends ESTestCase {
 
     private class TestHttpChannel implements HttpChannel {
         private final AtomicBoolean open = new AtomicBoolean(true);
-        private final AtomicReference<ActionListener<Void>> closeListener = new AtomicReference<>();
+        private final SubscribableListener<ActionListener<Void>> closeListener = new SubscribableListener<>();
         private final CountDownLatch closeLatch = new CountDownLatch(1);
 
         @Override
@@ -273,8 +275,7 @@ public class RestCancellableNodeClientTests extends ESTestCase {
         @Override
         public void close() {
             assertTrue("HttpChannel is already closed", open.compareAndSet(true, false));
-            ActionListener<Void> listener = closeListener.get();
-            if (listener != null) {
+            closeListener.andThenAccept(listener -> {
                 boolean failure = randomBoolean();
                 threadPool.generic().submit(() -> {
                     if (failure) {
@@ -284,11 +285,10 @@ public class RestCancellableNodeClientTests extends ESTestCase {
                     }
                     closeLatch.countDown();
                 });
-            }
+            });
         }
 
         private void awaitClose() throws InterruptedException {
-            assertNotNull("must set closeListener before calling awaitClose", closeListener.get());
             close();
             closeLatch.await();
         }
@@ -304,9 +304,8 @@ public class RestCancellableNodeClientTests extends ESTestCase {
             if (open.get() == false) {
                 listener.onResponse(null);
             } else {
-                if (closeListener.compareAndSet(null, listener) == false) {
-                    throw new AssertionError("close listener already set, only one is allowed!");
-                }
+                assertFalse("close listener already set, only one is allowed!", closeListener.isDone());
+                closeListener.onResponse(ActionListener.assertOnce(listener));
             }
         }
     }

